### PR TITLE
gh-108303: Move all doctest related files and tests to `Lib/test/test_doctest/`

### PR DESCRIPTION
**Pull request by sobolevn**
* Issue: gh-108303

### CODE REVIEW

File: Lib/test/support/pty_helper.py [hunk 1]
```diff
@@ -58,3 +58,23 @@ def terminate(proc):
...
+    A fake input stream for pdb'sx interactive debugger.  Whenever a
+    line is read, print it (to simulate the user typing it), and then
+    return it.  The set of lines to return is specified in the
+    constructor; they should not have trailing newlines.
+    """
+    def __init__(self, lines):
+        self.lines = lines
+
+    def readline(self):
+        line = self.lines.pop(0)
```
Comment (vstinner) [inline Lib/test/support/pty_helper.py:78]: Calling `list.pop(0)` frequently is inefficient. Maybe the constructor should create a `collections.deque()` and call `self.lines.popleft()` instead?

File: Makefile.pre.in [hunk 2]
```diff
@@ -2201,6 +2201,7 @@ TESTSUBDIRS=	idlelib/idle_test \
 		test/test_cppext \
 		test/test_ctypes \
 		test/test_dataclasses \
+		test/test_doctest \
```
Comment (vstinner) [inline Makefile.pre.in:2224]: Thank you for that one, ha ha ha!

File: Lib/test/support/pty_helper.py (hunk 1 above)
Comment (sobolevn) [reply Lib/test/support/pty_helper.py:78]: It is usually used in cases like `sys.stdin = FakeInput(['next', 'print(x)', 'continue'])`, I don't think it matters much. But, I can do this if you want :)

File: Lib/test/support/pty_helper.py [hunk 3]
```diff
@@ -58,3 +58,23 @@ def terminate(proc):
...
                         sel.modify(master, selectors.EVENT_READ)
+
+
+######################################################################
+## Fake stdin (for testing interactive debugging)
+######################################################################
+
+class FakeInput:
+    """
+    A fake input stream for pdb'sx interactive debugger.  Whenever a
```
Comment (brettcannon) [inline Lib/test/support/pty_helper.py]:
```suggestion
    A fake input stream for pdb's interactive debugger.  Whenever a
```

File: Lib/test/support/pty_helper.py [hunk 4]
```diff
@@ -58,3 +58,23 @@ def terminate(proc):
...
+    return it.  The set of lines to return is specified in the
+    constructor; they should not have trailing newlines.
+    """
+    def __init__(self, lines):
+        self.lines = lines
+
+    def readline(self):
+        line = self.lines.pop(0)
+        print(line)
+        return line+'\n'
```
Comment (brettcannon) [inline Lib/test/support/pty_helper.py]:
```suggestion
        return line + '\n'
```

**Review by brettcannon — APPROVED** [review]

**Comment by sobolevn** [comment]
I solved merge conflicts, cc @vstinner

**Comment by sobolevn** [comment]
> Since we are moving code, would it be worth it to split test_doctest.py which has 3 374 lines into multiple files?

I don't think so, because moving a file is a rather simple operation. It does not spoil `git blame`. While splitting files can be devastating to its history.

I work with `test_doctest` quite a lot, I don't think that it has maintability problems.

> About test_doctest2: why is it a separated file? Should it be renamed?

I agree that the naming of this file is not optimal. It also clashes with `test_doctest2.txt`, I don't think that this is intentional.

Looks like that all tests from it are just duplicates. We already test classes, classmethods, staticmethods and everything else in `test_doctest`. So, it can be simply removed.

It has some details that I want to explicitly port:
- `property` with a setter
- poluting namespace with other objects

**Review by vstinner — COMMENTED** [review]
I cannot see where Lib/test/test_doctest2.py test goes. Did you just remove it?

**Comment by sobolevn** [comment]
Yes:

> I agree that the naming of this file is not optimal. It also clashes with test_doctest2.txt, I don't think that this is intentional.

> Looks like that all tests from it are just duplicates. We already test classes, classmethods, staticmethods and everything else in test_doctest. So, it can be simply removed.

**Comment by vstinner** [comment]
> Looks like that all tests from it are just duplicates. We already test classes, classmethods, staticmethods and everything else in test_doctest. So, it can be simply removed.

I'm not convinced that test_doctest2 is useless. It calls `doctest.testmod(sys.modules[__name__])` and test_doctest2.py contains 2 non-ASCII docstrings. Example:

```py
"""...
We include some (random) encoded (utf-8) text in the text surrounding
the example.  It should be ignored:

ЉЊЈЁЂ
"""
```

It runs tests in verbose mode instead of running them in quiet mode:

* test_doctest: `doctest.testmod(unicodedata, verbose=False)` (quiet)
* test_doctest2: `doctest.testmod(sys.modules[__name__])` (verbose)

Even we agree that test_doctest2 is useless, I would prefer to *not* include this *silent* removal in the middle of such huge refactoring (move all tests code).

Please add back test_doctest2.py in the new test_doctest/ directory.

**Comment by vstinner** [comment]
Merged, thanks.

@sobolevn: The automated backport to 3.11 failed, do you want to attempt to backport the change to 3.11 manually?

**Comment by sobolevn** [comment]
Sure thing!

**Comment by brettcannon** [comment]
FYI this broke the WASI builds; https://buildbot.python.org/all/#/builders/1046/builds/4033 and https://buildbot.python.org/all/#/builders/1124/builds/623 .

**Comment by vstinner** [comment]
> FYI this broke the WASI builds; https://buildbot.python.org/all/#/builders/1046/builds/4033 and https://buildbot.python.org/all/#/builders/1124/builds/623 .

Fixed by https://github.com/python/cpython/commit/efb81a60f5ce7e192095230a0f7ff9684d6f835a of https://github.com/python/cpython/issues/114275

**Pull request merged**